### ISSUE TITLE
reduce communication bandwidth

### ISSUE THREAD

**Issue by luggi961** [OpenMined/PySyft#2934]
I implemented a distributed image classification using WebsocketWorkers and TrainConfig objects. I started with the pretrained MobileNetV2 model (downloaded with torchvision) ~8MB
While training, I inspected with WireShark how much traffic came across.

**Result:** 
~18MB in both directions each training round

**Assumption:** 
 The serialized message will be converted in a hexadecimal representation using the hexlify method (websocket_client.py). According to the documentation (https://docs.python.org/2/library/binascii.html) the resulting string is twice as long as the length of data. So I guess thats why the 8 MB model becomes 16 MB.

**Possible solution:**
Transmitting only the difference between old and new model would be quite more efficient

**Comment by iamtrask** [comment]
Mad props for taking the time to measure this.

**Comment by karlhigley** [comment]
If the differences were sparse, they'd be smaller. They often won't be though.

Gradient sparsification is a thing; maybe model delta sparsification could be too?

**Comment by karlhigley** [comment]
This seems like it might be more a matter of compression and encoding efficiency though? Base64 is supported by the `binascii` library and uses four characters for every three bytes (instead of two per byte), so that might be an easy change to make in order to improve bandwidth efficiency.

**Comment by vvmnnnkv** [comment]
In theory, websocket can handle binary. As least in JavaScript :) The trick is to avoid JSON because it can't contain binary.

**Comment by karlhigley** [comment]
Can we just remove the call to `hexlify` then?

**Comment by vvmnnnkv** [comment]
Needs checking why it was used, maybe there was a reason. I can say for grid.js, historically it uses JSON payloads, which forces us to send binary as base64

**Comment by karlhigley** [comment]
Sounds like base64 is a safe change to make then. Let's start there and then see if we can further optimize.

**Comment by vvmnnnkv** [comment]
Another kind of workaround specific to transfer learning case could be splitting the model into frozen/trianable parts. In transfer learning you typically don't want to retrain all weights.

**Comment by karlhigley** [comment]
Looks like use of `hexlify` was [introduced last March](https://github.com/OpenMined/PySyft/commit/6917846cf981485ca83f61142e073c9c343eba4c#diff-11214074701249affec97a37c2352924). There's no stated rationale in that PR; the title just suggests that the changes "fix bugs." Not sure if `hexlify` was part of those fixes or not.

**Comment by luggi961** [comment]
> What leads you to believe that transmitting the difference between the old and new model would be any smaller of a file? Would it compress better?

What I mean is an option where only the changed weights are transferred. Is this possible?
Correct me if I am wrong, but maybe this could be an approach.

```
diff = dict()
for idx, (federated_param, worker_param) in enumerate(zip(federated_model.parameters(), worker_model.parameters())):
        if federated_param.data.ne(worker_param.data).sum() > 0:
        diff[idx] = worker_param.data
```
